### ISSUE TITLE
Expose "powerflow" metrics: real time power import / export in watts

### ISSUE THREAD

**Issue by IsaacInsoll** [TimSoethout/goodwe-sems-home-assistant#92]
![IMG_9503](https://github.com/TimSoethout/goodwe-sems-home-assistant/assets/30609296/22dd5313-908b-4c18-87ca-58d862c94fc0)

I installed this integration yesterday and it’s great!
I can see energy produced by solar and the import and export kWh values. 

SEMS portal also has this real time power graph showing what is current being generated, whole home consumption and feed in/out in watts. 

I can see the realtime solar production, is there a way to see these other two values so I can make a real time power chart?

thanks 😀

**Comment by TimSoethout** [comment]
Good idea.
Maybe you can pull out the values using template sensors as mentioned in the readme?
Or wait on #91 that might also solve this.

**Comment by IsaacInsoll** [comment]
I didn’t think templates would work because when I inspect the HomeKit
device in the settings page I can’t see the raw power values there plus I
didn’t see much in terms of templates that measure in watts (not watt
hours) on the homepage.

I’ll have more of a play with it and see if I can work it out. Thank you

**Comment by TimSoethout** [comment]
Yes, it can only work if the value is present in the attributes of the inverter. Maybe we need to fetch it from another API call?

**Comment by IsaacInsoll** [comment]
Awesome. Is there anything I can do to help with that? Like provide a
certain log or something?

I’ve got a bit of programming experience but haven’t done anything with
SEMS or HA before

**Comment by TimSoethout** [comment]
You can try using the web portal, and looking at the web inspector for the network calls. Which API call contains the values you're looking for?

**Comment by IsaacInsoll** [comment]
Heres the UI:
![image](https://github.com/TimSoethout/goodwe-sems-home-assistant/assets/30609296/e02d3886-c506-4a3e-b30d-91c9f07498ee)
This is the request as a fetch:
```
fetch("https://au.semsportal.com/api/v2/PowerStation/GetPowerflow", {
  "headers": {
    "accept": "application/json, text/javascript, */*; q=0.01",
    "accept-language": "en-US,en;q=0.9",
    "content-type": "application/x-www-form-urlencoded; charset=UTF-8",
    "neutral": "0",
    "sec-ch-ua": "\"Google Chrome\";v=\"119\", \"Chromium\";v=\"119\", \"Not?A_Brand\";v=\"24\"",
    "sec-ch-ua-mobile": "?0",
    "sec-ch-ua-platform": "\"Windows\"",
    "sec-fetch-dest": "empty",
    "sec-fetch-mode": "cors",
    "sec-fetch-site": "same-site",
    "token": "<redacted-by-isaac>"
  },
  "referrer": "https://semsportal.com/",
  "referrerPolicy": "strict-origin-when-cross-origin",
  "body": "PowerStationId=<redacted-by-isaac>",
  "method": "POST",
  "mode": "cors",
  "credentials": "omit"
});
```
which returns:
```
{
  "language": "en",
  "function": null,
  "hasError": false,
  "msg": "success",
  "code": "0",
  "data": {
    "hasGenset": false,
    "hasMoreInverter": true,
    "hasPowerflow": true,
    "powerflow": {
      "pv": "5001W",
      "pvStatus": -1,
      "bettery": "0W",
      "betteryStatus": 0,
      "betteryStatusStr": null,
      "load": "825W",
      "loadStatus": -1,
      "grid": "4176W",
      "soc": 0,
      "socText": "0%",
      "hasEquipment": true,
      "gridStatus": -1,
      "isHomKit": false,
      "isBpuAndInverterNoBattery": false,
      "isMoreBettery": true,
      "genset": "0W",
      "gensetStatus": 0,
      "gridGensetStatus": 0
    },
    "hasGridLoad": true,
    "isParallelInventers": false,
    "isEvCharge": false,
    "evCharge": null
  },
  "components": {
    "para": "{\"model\":{\"PowerStationId\":\"<redacted-by-isaac>\"}}",
    "langVer": 195,
    "timeSpan": 24,
    "api": "http://au.semsportal.com:82/api/v2/PowerStation/GetPowerflow",
    "msgSocketAdr": null
  }
}
```
And the reason I think we need this is because I can't currently see these power values in the integration (i can only see kwh energy values):
![image](https://github.com/TimSoethout/goodwe-sems-home-assistant/assets/30609296/2f025e49-a2a9-422e-b635-0482ffa87fac)

So I'd love to be able to expose `grid` and `load` from `data.powerflow`. I'm not sure if `pv` matches exactly the value we already have for that or if it's different, but might as well expose that too so we have all 3 values to make this chart from the same 'source'

I'm not sure on how to make the changes as I've never done python before but I'm very happy to test this if you can make the changes. Or if you have some basic guidance on the correct way to implement this at a high level then I'm sure I can code something up over the next week.

**Comment by IsaacInsoll** [comment]
Hi, 
I just checked out your code to figure out which endpoint you are using, then I queried that endpoint to see if it already has the data we need. 
Good news, it's in there, so this should be an 'easy' job. 
The data is under `data.powerflow`: `load` and `grid` (and `pv` but i can already "see" that one in your integration)
Notably it's a string with `(W)` suffix rather than a 'real' number. 

<img width="198" alt="image" src="https://github.com/TimSoethout/goodwe-sems-home-assistant/assets/30609296/1ecadc16-e8c1-4642-81bf-250e2091d8eb">

**Comment by IsaacInsoll** [comment]
Looks like line 82 of `sensor.py` has a typo: 
```
powerflow["sn"] = result["homKit"]["sn"]
```
`homKit` should probably be `homeKit`

I'll also include a screenshot of docker logs, unfortunately it's in reverse chronological order
<img width="773" alt="image" src="https://github.com/TimSoethout/goodwe-sems-home-assistant/assets/30609296/cf2e1513-ca86-4d14-b2a5-da50ef0d414f">

**Comment by IsaacInsoll** [comment]
I noticed that I didn't have a serial number for homeKit so that was causing the third entity not to show up. 
In `sensor.py` I added a few lines:
```
           if hasPowerflow:
                if hasEnergeStatisticsCharts:
                    StatisticsCharts = { f"Charts_{key}": val for key, val in result["energeStatisticsCharts"].items() }
                    StatisticsTotals = { f"Totals_{key}": val for key, val in result["energeStatisticsTotals"].items() }
                    powerflow = { **result["powerflow"],  **StatisticsCharts, **StatisticsTotals }
                else:
                    powerflow = result["powerflow"]

                powerflow["sn"] = result["homKit"]["sn"]
                
                #isaac start
                if powerflow["sn"] is None:
                    powerflow["sn"] = "GW-HOMEKIT-NO-SERIAL"
                #isaac end
                
                #_LOGGER.debug("homeKit sn: %s", result["homKit"]["sn"])
                # This seems more accurate than the Chart_sum
                powerflow["all_time_generation"] = result["kpi"]["total_power"]

                data["homeKit"] = powerflow

```
directly edited on my live machine then restarted Home Assistant, and now the mystical third entity is there. 
So I'd say the real fault is incorrect handling of devices with no serial number. 
Just to confirm: this is the 'Smart Meter' not the 'HomeKit' so that's why it might not be registering a serial number even though it's got all the same data.

<img width="455" alt="image" src="https://github.com/TimSoethout/goodwe-sems-home-assistant/assets/30609296/2b1f6742-62c8-460a-a22e-e6f6149e9003">

I'm not going to make a PR because I just hacked this in and presume you might have a better way you want to handle it.

**Comment by TimSoethout** [comment]
Fixed by #115 
Released as https://github.com/TimSoethout/goodwe-sems-home-assistant/releases/tag/6.1.0
